### PR TITLE
Merge migration to query AST

### PR DESCRIPTION
**Pull request by MaceWindu**
**List of changes**

There were many fixes, that could look as not related to Merge, but they were blocking issues, discovered during Merge implementation.

- Merge query included into internal query AST
- `LinqService` support added for Merge queries
- Legacy `DataProvider.Merge API` ~removed. Check [migration notes](https://linq2db.github.io/articles/sql/merge/Merge-API-Migration.html), if you still use it~ implementation replaced with calls to main API
- added support for dynamic properties in Merge (except explicit target column setters)

- implemented missing `GetHashCode` overrides for AST classes with custom `Equals`
- fixed several cases when we loose `SqlDataType` column type information
- changed `ref` to `out` parameter for `MappingSchema.GetUnderlyingDataType(Type, ref bool)` method, as it is used only to return value from method
- AST visitors now throw exception, if node type is not handled, to detect missing visitor implementations. Added missing implementation for some nodes
- `ValueToSqlConverter.Convert()` will throw exception instead of converting value to SQL using `ToString()` call, if conversion is not defined, to help find missing conversions
- `SqlBuilder.BuildDataType()` method logic changed to first try to use `DbType`, and only if it is not set then call type name generation from `DataType` enumeration value (method still will throw if `DataType == Undefined`). Provider implementors need to move their custom type generation logic to new `BuildDataTypeFromDataType` method override.
- new virtual method `BuildTypedExpression` added to `SqlBuilder` to generate typed SQL expression. Default implementation generates `CAST({0} AS {TYPE})` and overriden for provides with other syntax
- fixed `double` and `float` values transfer by LinqService to use round-trip format
- added `BuildIdentityInsert` helper methods to SQL Server and Sybase SQL builders to generate `SET IDENTITY_INSERT <TABLE> ON/OFF` statements
- add `DateTime` to `DataType.DateTime` mapping for SQL Server
- enabled use of parameters for some providers where it was not enabled properly(Oracle, Informix, DB2)
- fixed bug when schema-specific converter promoted to default converters
- use culture-invariant formatting for primitives in LinqService
- implemented support for provider-specific data reader expressions in LinqService (#730)
- fix `StackOverflowException` in QueryVisitor.Find for recursive elements
- fix `SELECT * ...` queries generation for some cases
- fixed multiple issues and missing elements in Convert AST visitor
- remote context values de/serialization changed to use mapping schema instead of hardcoded logic in multiple places
- fixed missing SqlCe provider trimming of trailing spaces for nvarchar values

### CODE REVIEW

**Comment by MaceWindu** [comment]
**Tests status**
(core1 not tested, because we will remove it from 3.0)

| Provider \ Target | net46 | core2 | 
|-|-|-|
|SQLite.Classic| Pass ||
|SQLite.MS||
|Northwind.SQLite||
|TestNoopProvider|Pass|
|SqlCe||
|Access||
|DB2||
|Informix||
|SapHana||
|PostgreSQL||
|PostgreSQL.9.2||
|PostgreSQL.9.3||
|PostgreSQL.9.5||
|PostgreSQL.10||
|PostgreSQL.11||
|PostgreSQL.EDGE||
|MySql57||
|MariaDB||
|MySql||
|MySqlConnector||
|Northwind||
|SqlServer||
|SqlServer.2008||
|SqlServer.2008.1||
|SqlServer.2005||
|SqlServer.2005.1||
|SqlServer.2000||
|SqlServer.2012||
|SqlServer.2014||
|SqlServer.2017||
|SqlAzure.2012||
|Sybase||
|Sybase.Managed||
|Oracle.Managed||
|Oracle.Native||
|Firebird||
|Firebird3||

**Comment by MaceWindu** [comment]
image

File: Source/LinqToDB/Linq/Builder/MergeBuilder.On.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+{
+	internal partial class MergeBuilder
+	{
+		internal class On : MethodCallBuilder
+		{
+			protected override bool CanBuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
+			{
+				return methodCall.Method.IsGenericMethod
+					&& (LinqExtensions.OnMethodInfo1        .GetGenericMethodDefinition() == methodCall.Method.GetGenericMethodDefinition()
+					 || LinqExtensions.OnMethodInfo2        .GetGenericMethodDefinition() == methodCall.Method.GetGenericMethodDefinition()
```
Comment (sdanyliv) [inline Source/LinqToDB/Linq/Builder/MergeBuilder.On.cs]: GetGenericMethodDefinition can be called once during LinqExtensions initialization. Same for other builders.

File: Source/LinqToDB/ServiceModel/ServiceModelDataReader.cs [hunk 2]
```diff
@@ -162,98 +59,82 @@ public Type GetFieldType(int i)
...
 		}
 
-		public object this[string name]
-		{
-			get { return GetValue(GetOrdinal(name)); }
-		}
+		bool IDataReader.IsClosed              => throw new NotImplementedException();
+		int IDataReader.RecordsAffected        => throw new NotImplementedException();
+		DataTable IDataReader.GetSchemaTable() => throw new NotImplementedException();
+		bool IDataReader.NextResult()          => throw new NotImplementedException();
```
Comment (sdanyliv) [inline Source/LinqToDB/ServiceModel/ServiceModelDataReader.cs]: Formatting

File: Source/LinqToDB/SqlQuery/SqlMergeOperationClause.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+		internal SqlMergeOperationClause(
+			MergeOperationType type,
+			SqlSearchCondition where,
+			SqlSearchCondition whereDelete,
+			IEnumerable<SqlSetExpression> items)
+		{
+			OperationType = type;
+			Where = where;
+			WhereDelete = whereDelete;
```
Comment (sdanyliv) [inline Source/LinqToDB/SqlQuery/SqlMergeOperationClause.cs]: Formatting

File: Source/LinqToDB/Data/DataConnectionExtensions.cs [hunk 4]
```diff
@@ -1478,523 +1478,5 @@ public static BulkCopyRowsCopied BulkCopy<T>(this ITable<T> table, IEnumerable<T
 		}
 
 		#endregion
-
-		#region Merge
```
Comment (igor-tkachev) [inline Source/LinqToDB/Data/DataConnectionExtensions.cs]: Where is that old API?

**Comment by MaceWindu** [comment]
@it old API restored

**Pull request merged**